DDD-7.md

```diff
@@ -0,0 +1,421 @@
+* `INITIALIZING` - switch to this state at the very beginning of the `run()` method, engine is in this state during initializing of the connector, starting the connector itself and calling `DebeziumEngine.ConnectorCallback.connectorStarted()` callback,
+* `CREATING_TASKS` - switch to this state after successful start of the connector, configurations of the tasks are being created and initialized,
+* `STARING_TASKS` - tasks are being started, each in separate thread; stays in this stage until tasks are started, start of the tasks have failed or time specified by `TASK_MANAGEMENT_TIMEOUT_MS` option elapsed,
+* `POLLING_TASKS` - tasks polling has started; this is the main phase when the data are produced and engine stays in this stage until it starts to shut down or exception was thrown,
+* `STOPPING` - the engine is being stopped, either because engine's `close()` method was called or an exception was thrown; offsets are stored, `ExecutorService` for processing records shut down,  tasks and connector are stopped in this stage,
+* `STOPPED` - engine has been stopped; final state, cannot move any further from this state and any call on engine object in this state should fail
+
+Possible state transitions:
+
+* `STARTING` -> `INITIALIZING`
+* `INITIALIZING` -> `CREATING_TASKS`
+* `CREATING_TASKS` -> `STARING_TASKS`
+* `STARING_TASKS` -> `POLLING_TASKS`
+* (`STARTING` | `INITIALIZING` | `CREATING_TASKS` | `STARING_TASKS` | `POLLING_TASKS`) -> `STOPPING` 
+* `STOPPING`  -> `STOPPED` 
+
+#### Preventing resource leakage
+
+An engine stage that requires special attention is the one during which tasks are started.
+At this stage connections the databases are being created and if something bad happens or the  engine is shut down while tasks are being started, it may result in various leaked resources, e.g. unclosed replication slots.
+For more detail about possible issues, please see [DBZ-2534](https://issues.redhat.com/browse/DBZ-2534).
+To prevent this situation, transition from `STARING_TASKS` into  `STOPPING`  won't be possible by calling engine's `close()` method.
+Also, `STARING_TASKS` must finish completely.
+Even if one of the threads fails to start the task it was running, the main (engine) thread has to wait for all other tasks until they finish (no matter if successfully or not) before moving into the `STOPPING` state.
+In general, transition `STARING_TASKS` ->  `STOPPING` is possible, but it should happen only in the case when an exception was thrown from method starting tasks and only until all threads starting tasks have finished.
+
+#### Exception handling
+
+Retriable exceptions are handled at the place where they happen and relevant action is retried until `ERRORS_MAX_RETRIES` attempts are exhausted.
+Contrary to the existing `EmebeddedEngine` implementation, task is not restarted at this point (**TODO**: *re-think, why is the task restarted in the current implementation?*).
```

Comment:
Restart loop was added into embedded engine in [11c588fb0](https://github.com/debezium/debezium/commit/11c588fb0350495fe8eb5a085ac9791b1b1f4c53#diff-e814dc964227048cece0a87ca5a8290f86c6d3a07cf7080c3dc880006dad1ab9R841-R866). Justification in related [DBZ-4629](https://issues.redhat.com/browse/DBZ-4629) is not very clear, but as per [this comment](https://github.com/debezium/debezium/pull/3857/files#r990926825), it seems that the main reason was that some connectors thrown exceptions from their `start()` method which weren't retriable. We have most of the exception retriable now (since [DBZ-5244](https://issues.redhat.com/browse/DBZ-5244)), but [this comment](https://github.com/debezium/debezium/pull/3857#issuecomment-1245313387) claims it doesn't solve the issue. It's not clear to me which issue exactly it doesn't solve, but as we stop and restart the task in `poll()` method of `BaseSourceTask`, I don't believe we need to do it in engine again (currently, retries are applied only for exception thrown from `task.poll()` call). I believe we just need to catch `RetriableException` thrown from `BaseSourceTask::poll()` method and let the engine to call `BaseSourceTask::poll()` again, which should result into calling `BaseSourceTask::startIfNeededAndPossible()` which should eventually start the task again. If there is any issue with it, it should be IMHO solved in the right place where the issue actually is and adding additional task stop/start in embedded engine shouldn't be needed.